File: packages/apps/console/src/cloud/contract.ts
export interface CloudState {
  readonly signedIn: boolean;
  readonly credits: number;
  readonly gateway: string | null;
}

export interface CloudCapability {
  state(): CloudState;
}

let cloud: CloudCapability | undefined;

export function registerCloud(capability: CloudCapability): void {
  cloud = capability;
}

export function getCloud(): CloudCapability {
  if (cloud === undefined) throw new Error("Cloud capability is not registered");
  return cloud;
}

File: mobile/lib/cloud/contract.dart
class CloudState {
  const CloudState({required this.signedIn, required this.credits, required this.gateway});

  final bool signedIn;
  final int credits;
  final Uri? gateway;
}

abstract interface class CloudCapability {
  CloudState state();
}

CloudCapability? _cloud;

void registerCloud(CloudCapability capability) {
  _cloud = capability;
}

CloudCapability get cloud {
  final capability = _cloud;
  if (capability == null) throw StateError('Cloud capability is not registered');
  return capability;
}
